File: Frontend/src/api/identity.ts
import { http } from '../utils/http'

// 与 Backend/app/schemas 对齐的 TS 类型

/** 账号简要信息 */
export interface AccountBrief {
  id: number
  is_temporary: boolean
  username: string | null
  role: string
}

/** 玩家简要信息 */
export interface PlayerBrief {
  uuid: string
  name: string
  role: string
}

/** 认证响应（exchange / login / register / claim-bind 共用 shape） */
export interface AuthResponse {
  access_token: string
  refresh_token: string
  token_type: 'Bearer'
  player: PlayerBrief
  account: AccountBrief
}

/** /me 响应 */
export interface MeResponse {
  account: AccountBrief
  players: PlayerBrief[]
  active_uuid: string
}

/** /web-accounts/me 响应 */
export interface MyAccountResponse {
  account: AccountBrief
  players: PlayerBrief[]
}

/** 绑定短码响应 */
export interface BindCodeResponse {
  short_code: string
  expires_in: number
}

/** 绑定确认响应 */
export interface BindResultResponse {
  player: PlayerBrief
  account: AccountBrief
}

/** POST /auth/exchange —— 一次性 token 换 JWT pair */
export async function exchangeToken(token: string): Promise<AuthResponse> {
  const { data } = await http.post<AuthResponse>('/auth/exchange', { token })
  return data
}

/** POST /auth/login —— 用户名密码登录 */
export async function passwordLogin(username: string, password: string): Promise<AuthResponse> {
  const { data } = await http.post<AuthResponse>('/auth/login', { username, password })
  return data
}

/** GET /me —— 取当前身份信息（账号 + 绑定的玩家列表 + 当前激活 uuid） */
export async function fetchMe(): Promise<MeResponse> {
  const { data } = await http.get<MeResponse>('/me')
  return data
}

/** POST /web-accounts/register —— 注册永久账号（临时会话转正） */
export async function register(username: string, password: string): Promise<AuthResponse> {
  const { data } = await http.post<AuthResponse>('/web-accounts/register', { username, password })
  return data
}

/** GET /web-accounts/me —— 取账号详情（含绑定玩家列表） */
export async function getMyAccount(): Promise<MyAccountResponse> {
  const { data } = await http.get<MyAccountResponse>('/web-accounts/me')
  return data
}

/** POST /bind/issue —— 申请绑定短码 */
export async function issueBindCode(): Promise<BindCodeResponse> {
  const { data } = await http.post<BindCodeResponse>('/bind/issue')
  return data
}

/** POST /bind/confirm —— 确认绑定（输入游戏内 !!PCH bind 给的短码） */
export async function confirmBind(shortCode: string): Promise<BindResultResponse> {
  const { data } = await http.post<BindResultResponse>('/bind/confirm', { short_code: shortCode })
  return data
}

/** POST /bind/claim —— 临时会话下绑定到已有永久账号 */
export async function claimBind(username: string, password: string): Promise<AuthResponse> {
  const { data } = await http.post<AuthResponse>('/bind/claim', { username, password })
  return data
}

File: Backend/app/schemas/identity.py
"""身份管理 API 请求/响应模型。

与前端 Frontend/src/api/identity.ts 类型对齐：
- AccountBrief / PlayerBrief / AuthResponse=TokenExchangeResponse / MeResponse / MyAccountResponse
  / BindCodeResponse=BindTokenIssueResponse / BindResultResponse=BindConfirmResponse
"""
import re
import uuid

from pydantic import BaseModel, Field, field_validator


def _validate_username(v: str) -> str:
    from app.core.config import get_settings

    s = get_settings()
    if not (s.username_min_length <= len(v) <= s.username_max_length):
        raise ValueError(
            f"username length must be between {s.username_min_length} and {s.username_max_length}"
        )
    if not re.match(r"^[A-Za-z0-9_-]+$", v):
        raise ValueError("username may only contain letters, digits, _ and -")
    return v


def _validate_password(v: str) -> str:
    from app.core.config import get_settings

    s = get_settings()
    if not (s.password_min_length <= len(v) <= s.password_max_length):
        raise ValueError(
            f"password length must be between {s.password_min_length} and {s.password_max_length}"
        )
    return v


# ===== 请求模型 =====


class PasswordLoginRequest(BaseModel):
    username: str
    password: str


class RegisterRequest(BaseModel):
    username: str
    password: str

    @field_validator("username")
    @classmethod
    def validate_username(cls, v: str) -> str:
        return _validate_username(v)

    @field_validator("password")
    @classmethod
    def validate_password(cls, v: str) -> str:
        return _validate_password(v)


class ClaimBindRequest(BaseModel):
    username: str
    password: str

    @field_validator("username")
    @classmethod
    def validate_username(cls, v: str) -> str:
        return _validate_username(v)

    @field_validator("password")
    @classmethod
    def validate_password(cls, v: str) -> str:
        return _validate_password(v)


class BindTokenRequest(BaseModel):
    """POST /bind/token 请求体（MCDR bind_client.request_bind_token POST JSON body {uuid, name}）。"""

    uuid: uuid.UUID
    name: str


class BindConsumeRequest(BaseModel):
    """POST /bind/consume 请求体（仅 short_code；UUID 走 X-Player-UUID header）。"""

    short_code: str


class BindConfirmRequest(BaseModel):
    short_code: str


class UpdateDisplayNameRequest(BaseModel):
    """PATCH /web-accounts/me 请求体：设置自定义昵称（sheets 三端显示名主源）。

    空白 strip；schema ``min_length=1`` 拒纯空白（与迁移 CHECK 一致）。
    """

    display_name: str = Field(min_length=1, max_length=64)


# ===== 响应模型（引用 schemas.auth 的 AccountBrief/PlayerBrief 保持单一来源） =====


class BindTokenIssueResponse(BaseModel):
    """绑定短码响应（前端 BindCodeResponse）。"""
    short_code: str
    expires_in: int


class BindConfirmResponse(BaseModel):
    """POST /bind/confirm 响应（前端 BindResultResponse）。

    契约：绑定不改 account，前端继续用现有 JWT；响应只含 player + account，不含 token。
    """
    player: "PlayerBrief"
    account: "AccountBrief"


class BindConsumeResponse(BaseModel):
    """POST /bind/consume 响应（MCDR bind_client.consume_bind_code 依赖）。

    契约：{status, account, player} —— MCDR 客户端用 account+player 渲染回执。
    """
    status: str
    player: "PlayerBrief"
    account: "AccountBrief"


class MyAccountResponse(BaseModel):
    """GET /web-accounts/me 响应（前端 MyAccountResponse）。

    契约：{account: AccountBrief, players: [PlayerBrief]}。
    """
    account: "AccountBrief"
    players: list["PlayerBrief"]


# 延迟引用解析（避免与 schemas.auth 循环 import）
from app.schemas.auth import AccountBrief, PlayerBrief  # noqa: E402

BindConfirmResponse.model_rebuild()
BindConsumeResponse.model_rebuild()
MyAccountResponse.model_rebuild()
